src/metabase/search/ingestion.clj

```
@@ -115,25 +115,43 @@
   false)
 
 (defn update!
-  "Update all active engines' existing indexes with the given documents"
-  [documents-reducible]
+  "Update all active engines' existing indexes with the given documents. Passed remove-documents will be deleted from the index."
+  [documents-reducible remove-documents]
   (when-let [engines (seq (search.engine/active-engines))]
     (if (= 1 (count engines))
-      (search.engine/update! (first engines) documents-reducible)
-      ;; TODO um, multiplexing over the reducible awkwardly feels strange. We at least use a magic number for now.
-      (doseq [batch (eduction (partition-all 150) documents-reducible)
-              e     engines]
-        (search.engine/update! e batch)))))
+      (u/prog1 (search.engine/update! (first engines) documents-reducible)
+        (doseq [[group ids] (u/group-by first second remove-documents)]
+          (search.engine/delete! (first engines) group ids)))
+      (u/prog1
+        ;; TODO um, multiplexing over the reducible awkwardly feels strange. We at least use a magic number for now.
+        (doseq [batch (eduction (partition-all 150) documents-reducible)
+                e engines]
+          (search.engine/update! e batch))
+        (doseq [[group ids] (u/group-by first second remove-documents)]
+          (search.engine/delete! (first engines) group ids))))))
+
+(defn- extract-model-and-id [update]
+  (let [[model value] update]
+    [model (str (cond
+                  (= := (first value))
+                  (first (filter integer? value))
+
+                  (= :and (first value))
+                  (first (first (filter not-empty (map (partial filter integer?) (rest value)))))))]))
```

1. can you make this a case statement, a bit more terse but more importantly will fail on unhandled expressions
2. this looks broken for compound ids? take a look at the [indexed-entity](https://github.com/metabase/metabase/blob/520f30b3f3de4acd8dc9b78876215caad83cb951/src/metabase/indexed_entities/models/model_index.clj#L187).

I see that model is using a direct sql expression, which is firstly not great for when we want to support nosql search indexes like elastic, but also not great for leveraging the spec as data for how to "decompile" this sql expression. I think refactoring it to just be defined as just `[:key1 :key2 :key3]` in the spec and converted to sql elsewhere could make it easier to write generic code here.

Lastly we shouldn't have "garbage" in garbage out semantics for non-usable input, we should either throw an error or maybe this a no-op, but I'm not sure which, because I'm not sure if there are any such expected cases.